### PR TITLE
Fix a bug where a DNS resolution is timed out when IPv6 is enabled

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

If IPv6 is enabled, the default DNS resolver sends both `A` and `AAAA` record types to get available IP addresses. `DefaultDnsResolver` waits for all questions to be resolved.
https://github.com/line/armeria/blob/94a1bf40ff41e1cfa9cd0c1a08f7bde14a7153c9/core/src/main/java/com/linecorp/armeria/internal/client/dns/DefaultDnsResolver.java#L126-L131

It is common that IPv6 record is not registered for servers that are used for internal communication. If a DNS server does not have an IPv6 address for a DNS query, the default DNS resolver continues to look up the next search domains.

Even if a DNS resolver acquired an IPv4 address successfully, `DefaultDnsResolver` keeps waiting for an IPv6 response. It would be a long-running task to check all search domains.

For that case, users frequently see `DnsTimeoutException` without `-Djava.net.preferIPv4Stack=true` which sends a DNS question only for IPv4.

Modifications:

- Fix `DefaultDnsResolver` to make a DNS query successful immediately if the first preferred records are found.
  - `DefaultDnsResoluver` does not wait for all questions to be resolved.
  - If a less preferred response is received early, the resolver will wait for the highly preferred one until the query is timed out.
  - If resolved records exist when a timeout is triggered, use the value.

Result:

You no longer see `DnsTimeoutException` when an IPv6 network interface is enabled.

### CODE REVIEW

**Review by trustin — APPROVED** [review]
:+1: Thanks for fixing this.

File: core/src/test/java/com/linecorp/armeria/internal/client/dns/DefaultDnsResolverTest.java [hunk 1]
```diff
@@ -92,4 +111,152 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
             resolver.close();
         }
     }
+
+    @CsvSource({ "IPV4_PREFERRED", "IPV6_PREFERRED" })
```
Comment (jrhee17) [inline core/src/test/java/com/linecorp/armeria/internal/client/dns/DefaultDnsResolverTest.java:115]: Didn't know we could use `CsvSource` with enum types 😄

**Review by jrhee17 — APPROVED** [review]
Looks really nice 👍    Thanks for debugging/fixing this issue @ikhoon ! 🙇 👍 🙇

File: core/src/main/java/com/linecorp/armeria/internal/client/dns/DefaultDnsResolver.java [hunk 2]
```diff
@@ -170,6 +137,52 @@ CompletableFuture<List<DnsRecord>> resolveAll(DnsQuestionContext ctx, List<? ext
...
+        }
+
+        if (cause != null) {
+            results[order] = Exceptions.peel(cause);
+        } else {
+            results[order] = records;
+        }
+
+        for (int i = 0; i < results.length; i++) {
+            final Object result = results[i];
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/internal/client/dns/DefaultDnsResolver.java:158]: nit:
```suggestion
        for (Object result : results) {
```

File: core/src/main/java/com/linecorp/armeria/internal/client/dns/DefaultDnsResolver.java [hunk 3]
```diff
@@ -170,6 +137,52 @@ CompletableFuture<List<DnsRecord>> resolveAll(DnsQuestionContext ctx, List<? ext
...
+                // A highly preferred question hasn't finished yet.
+                return;
+            }
+
+            if (result instanceof Throwable) {
+                // Skip a failed question and look up low priority ones.
+                continue;
+            }
+
+            if (result instanceof List) {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/internal/client/dns/DefaultDnsResolver.java:169]: `assert result instanceof List;` would be more clear?

**Review by minwoox — APPROVED** [review]
Left minor comments. Thanks for fixing this!

**Comment by ikhoon** [comment]
Thanks for reviewing!

File: core/src/main/java/com/linecorp/armeria/internal/client/dns/DefaultDnsResolver.java [hunk 4]
```diff
@@ -116,50 +104,29 @@ private CompletableFuture<List<DnsRecord>> resolveOne(DnsQuestionContext ctx, Dn
...
-                        }
-                    }
-                    return Exceptions.throwUnsafely(cause);
-                });
+        final CompletableFuture<List<DnsRecord>> future = new CompletableFuture<>();
+        final Object[] results = new Object[questions.size()];
+        for (int i = 0; i < questions.size(); i++) {
+            final int order = i;
+            delegate.resolve(ctx, questions.get(i)).handle((records, cause) -> {
+                assert executor.inEventLoop();
```
Comment (jrhee17) [inline core/src/main/java/com/linecorp/armeria/internal/client/dns/DefaultDnsResolver.java:112]: While reviewing #4748, i'm wondering is it safe to assume this?
If `delegate` is a `CachingDnsResolver`, I guess this assertion might not be true

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/internal/client/dns/DefaultDnsResolver.java:112]: Your analysis is correct.
I assumed an event loop thread invokes `DefaultDnsResolver` which turned out to be false.

Comment (jrhee17) [reply core/src/main/java/com/linecorp/armeria/internal/client/dns/DefaultDnsResolver.java:112]: I actually meant that `CachingDnsResolver` is a cache which contains `CompletableFuture`s.
Depending on which `eventloop` completes the `CompletableFuture`, the original `eventloop` doesn't necessarily handle the callback

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/internal/client/dns/DefaultDnsResolver.java:112]: The `CompletableFuture`s from `CachingDnsResolver` are all completed by the same `eventLoop` so isn't it okay?

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/internal/client/dns/DefaultDnsResolver.java:112]: A `CachingDnsResolver` is executed by a `DefaultDnsResolver` that is executed by an event loop. Therefore, one event loop accesses `CachingDnsResolver`.

**Pull request merged**